volatility3/framework/__init__.py

```diff
@@ -57,20 +59,64 @@ class Deprecation:
             additional_information: Information appended at the end of the deprecation message
         """
 
         def decorator(deprecated_func):
             @functools.wraps(deprecated_func)
             def wrapper(*args, **kwargs):
-                nonlocal replacement, additional_information
-                deprecation_msg = f"Method \"{deprecated_func.__module__ + '.' + deprecated_func.__name__}\" is deprecated, use \"{replacement.__module__ + '.' + replacement.__name__}\" instead. {additional_information}"
-                vollog.warning(deprecation_msg)
+                nonlocal replacement, replacement_base_class_required_version, additional_information
+                # Prevent version mismatches between deprecated (proxy) methods and the ones they proxy
+                if replacement_base_class_required_version is not None and callable(
+                    replacement
+                ):
+                    # example: replacement = volatility3.MyClass.my_dummy_function
+                    # "MyClass.my_dummy_function" -> "MyClass"
+                    replacement_base_class_name = replacement.__qualname__.split(".")[0]
+                    # replacement.__globals__ example: {'MyClass': <class 'volatility3.MyClass'>}
+                    replacement_base_class = replacement.__globals__.get(
+                        replacement_base_class_name
+                    )
+
+                    # Verify that the base class inherits from VersionableInterface
+                    if inspect.isclass(replacement_base_class) and issubclass(
+                        replacement_base_class,
+                        interfaces.configuration.VersionableInterface,
+                    ):
+                        # Construct a requirement
+                        req = requirements.VersionRequirement(
+                            name=replacement_base_class.__name__,
+                            component=replacement_base_class,
+                            version=replacement_base_class_required_version,
+                        )
+                        # Verify the requirement
+                        if not req.matches_required(
+                            req._version, req._component.version
+                        ):
+                            full_unsat_req_path = (
+                                deprecated_func.__module__
+                                + "."
+                                + deprecated_func.__qualname__
+                                + "."
+                                + req.name
+                            )
+                            # Catched by the cli and redirected to process_unsatisfied_exceptions
+                            raise exceptions.UnsatisfiedException(
```

Comment:
I really wouldn't worry about making the req for this.  It's going to be an internal issue for the codebase, so it just has to alert the user that something went very wrong and they should file a bug about it (including the original function called, and the proxied class it tried to call, both of which we should know).